writer.go

```diff
@@ -18,142 +32,142 @@ type Writer struct {
 	}
 
-	sdk.Logger(ctx).Trace().Msg("upserted record")
-	return nil
+	return int(upserted), nil
 }
 
-// Delete deletes records by a key.
-func (w *Writer) Delete(ctx context.Context, record sdk.Record) error {
-	ID := []string{recordID(record.Key)}
-
-	err := w.index.DeleteVectorsById(&ctx, ID)
+func (w *Writer) DeleteVectorsByID(ctx context.Context, vectorIDs []string) error {
+	err := w.index.DeleteVectorsById(&ctx, vectorIDs)
 	if err != nil {
-		return fmt.Errorf("\n error deleting record: %v ", err)
+		return fmt.Errorf("error deleting record: %w", err)
 	}
 
-	sdk.Logger(ctx).Trace().Msg("Successful record delete.")
 	return nil
 }
 
 func (w *Writer) Close() error {
-	return w.index.Close()
-}
-
-// NewPineconeClient takes the Pinecone Index URL string in Config and splits into respective parts to establish a
-// connection
-func NewPineconeClient(ctx context.Context, config DestinationConfig) (*pinecone.Client, *pinecone.IndexConnection, error) {
-	sdk.Logger(ctx).Trace().Msg("Creating a Pinecone Client.")
-
-	client, err := pinecone.NewClient(pinecone.NewClientParams{
-		ApiKey: config.PineconeAPIKey,
-	})
-	if err != nil {
-		return nil, nil, fmt.Errorf("error creating Pinecone client: %v", err)
+	if err := w.index.Close(); err != nil {
+		return fmt.Errorf("error closing writer: %w", err)
 	}
-	sdk.Logger(ctx).Info().Msg("created pinecone client")
-
-	// index urls should have their protocol trimmed
-	hostURL := strings.TrimPrefix(config.PineconeHostURL, "https://")
 
-	index, err := client.Index(hostURL)
-	if err != nil {
-		return nil, nil, fmt.Errorf("error establishing index connection: %v", err)
-	}
-	sdk.Logger(ctx).Info().Msgf("created pinecone index")
+	return nil
+}
 
-	return client, index, nil
+func recordID(key sdk.Data) string {
+	return string(key.Bytes())
 }
 
-func recordID(Key sdk.Data) string {
-	key := Key.Bytes()
-	return string(key)
+type sparseValues struct {
+	Indices []uint32  `json:"indices"`
+	Values  []float32 `json:"values"`
 }
 
 type recordPayload struct {
+	ID           string       `json:"id"`
 	Values       []float32    `json:"values"`
-	SparseValues sparseValues `json:"sparse_values"`
+	SparseValues sparseValues `json:"sparse_values,omitempty"`
 }
 
 func (r recordPayload) PineconeSparseValues() *pinecone.SparseValues {
-	v := &pinecone.SparseValues{r.SparseValues.Indices, r.SparseValues.Values}
+	// the used pinecone go client needs a nil pointer when no sparse values given, or else it
+	// will throw a "Sparse vector must contain at least one value" error
+	if len(r.SparseValues.Indices) == 0 && len(r.SparseValues.Values) == 0 {
+		return nil
+	}
+
+	v := &pinecone.SparseValues{
+		Indices: r.SparseValues.Indices,
+		Values:  r.SparseValues.Values,
+	}
 	return v
 }
 
-type sparseValues struct {
-	Indices []uint32  `json:"indices"`
-	Values  []float32 `json:"values"`
+func parseVector(rec sdk.Record) (*pinecone.Vector, error) {
+	id := recordID(rec.Key)
+
+	payload, err := parseRecordPayload(rec)
```

Comment:
Maybe something like `parsePineconeValues`?

Reply:
Same as above, I still see the old name, but the conversation was resolved. Maybe you need to push a commit?

Reply:
Yeah, did not push the commit, sorry!